crates/feldera-storage/src/file/format.rs

```diff
@@ -0,0 +1,386 @@
+//!   [`DataBlockHeader::value_map_varint`].
+//!
+//! * The "row groups", which point from a row in this column to the associated
+//!   rows in the next column.  The offset to the value map is specified in
+//!   [`DataBlockHeader::row_groups_ofs`] and the format in
+//!   [`DataBlockHeader::row_group_varint`].  Data blocks in the last column do
+//!   not have row groups.
+//!
+//! # Index blocks
+//!
+//! An index block consists of the following, in order:
+//!
+//! * [`IndexBlockHeader`].
+//!
+//! * A sequence of bounds serialized with [`rkyv`], one pair for each of
+//!   [`IndexBlockHeader::n_children`].  The first value in each pair is the
+//!   smallest value in the child tree and the second is the largest value.
+//!
+//! * A "bound map", which holds a byte offset from the start of the block to
+//!   the start of each bound.  The offset to the bound map is specified in
+//!   [`IndexBlockHeader::bound_map_offset`] and the format in
+//!   [`IndexBlockHeader::bound_map_varint`].
+//!
+//! * An array of "row totals", one for each of
+//!   [`IndexBlockHeader::n_children`].  The first row total is the total number
+//!   of rows in the first child tree, the second row total is that plus the
+//!   total number of rows in the second child tree, and so on.
+//!
+//! * An array of "child pointers", one for each of
+//!   [`IndexBlockHeader::n_children`].  Each one of these points to a child
+//!   block.  For `size` and `offset` both in bytes, each pointer is encoded as
+//!   `(offset >> 7) | size.trailing_zeros()`, which allows `size` to range up
+//!   to `2**31` bytes.
+
+use binrw::{binrw, BinRead, BinResult, BinWrite, Error as BinError};
+#[cfg(doc)]
+use crc32c;
+use num_derive::FromPrimitive;
+use num_traits::FromPrimitive;
+use rkyv::{with::Inline, AlignedVec, Archive, Serialize};
+
+use super::Rkyv;
+
+/// Increment this on each incompatible change.
+pub const VERSION_NUMBER: u32 = 1;
+
+/// 8-byte header at the beginning of each block.
+///
+/// A block does not identify its own size, so any reference to a block must
+/// also include the block's size.
+#[binrw]
+#[derive(Debug)]
+pub struct BlockHeader {
+    /// 32-bit [`crc32c`] checksum of the remainder of the block.
+    pub checksum: u32,
+
+    /// Magic number.  Magic numbers begin with `LF`, which stands for "layer
+    /// file".
+    pub magic: [u8; 4],
+}
+
+impl BlockHeader {
+    pub(crate) fn new(magic: &[u8; 4]) -> Self {
+        Self {
+            checksum: 0,
+            magic: *magic,
+        }
+    }
+}
+
+/// File trailer block.
+///
+/// Padded with zeros to exactly fill a 4-kB block.
+///
+/// Serialized and deserialize automatically with [`mod@binrw`].
+#[binrw]
+#[derive(Debug)]
+pub struct FileTrailer {
+    /// Block header with "LFFT" magic.
+    #[brw(assert(&header.magic == b"LFFT", "file trailer has bad magic"))]
+    pub header: BlockHeader,
+
+    /// Must be [`VERSION_NUMBER`].
+    pub version: u32,
```

Comment:
It should be possible to support multiple versions. This comment is meant to state the current constraint, which is there because there is only one version so far.

I'll make the comment more detailed.